feat: Implement Indexed builder class

zonca opened the pull request:
I compare the implementation to a branch where I have just copy-pasted `IndexedOption` to more easily compare the differences between the 2 implementations.

zonca committed:
feat: remove handling of invalid values from copy of IndexedOption

ManasviGoyal commented on header-only/layout-builder/awkward/LayoutBuilder.h:
```diff
@@ -1088,7 +1088,6 @@ namespace awkward {
     /// @brief Builds an IndexedArray which consists of an `index` buffer.
-    /// The negative values in the index are interpreted as missing.
     ///
     /// The index values can be 64-bit signed integers `int64`, 32-bit signed
     /// integers `int32`.
```
```suggestion
    /// The index values can be 64-bit signed integers `int64`, 32-bit signed
    /// integers `int32`.
```
The valid types are - 64-bit signed integers `int64`, 32-bit signed integers `int32` or 32-bit unsigned integers `uint32`.

ManasviGoyal commented on header-only/layout-builder/awkward/LayoutBuilder.h:
```diff
@@ -1201,22 +1179,6 @@ namespace awkward {
         return index_.length();
       }
 
-      /// @brief Checks for validity and consistency.
-      bool
-      is_valid(std::string& error) const noexcept {
-        if (content_.length() != last_valid_ + 1) {
-          std::stringstream out;
-          out << "Indexed node" << id_ << " has content length "
-              << content_.length() << " but last valid index is " << last_valid_
-              << "\n";
-          error.append(out.str());
-
-          return false;
-        } else {
-          return content_.is_valid(error);
-        }
-      }
```
We still need an `is_valid` method. We can compare the `content_.length()` and `index_.length()` for this case.

ManasviGoyal commented on header-only/layout-builder/awkward/LayoutBuilder.h:
```diff
@@ -1129,43 +1126,25 @@ namespace awkward {
-      append_valid() noexcept {
-        last_valid_ = content_.length();
-        index_.append(last_valid_);
+      append() noexcept {
```
I think we would want the method name to be different than what we have for `Numpy`. It can be `append_index` maybe.

ManasviGoyal commented on header-only/layout-builder/awkward/LayoutBuilder.h:
```diff
@@ -1129,43 +1126,25 @@ namespace awkward {
       /// Just an interface; not actually faster than calling append many times.
       BUILDER&
-      extend_valid(size_t size) noexcept {
+      extend(size_t size) noexcept {
```
Similarly, this can be `extend_index`.

ManasviGoyal commented:
@zonca The PR name should have a `feat:` to make the Lint PR test pass.

zonca committed:
doc: support also uint32

zonca committed:
feat: implement is_valid checking index and content lengths

zonca committed:
feat: renamed append and extend to append_index and extend_index

zonca committed:
test: implemented test_Indexed, should fail

zonca commented on header-only/tests/test_1494-layout-builder.cpp:
```diff
@@ -1209,6 +1212,62 @@ test_Regular_size0() {
+
+  builder.extend_index(1);
+  subbuilder.extend(data, 3);
+
```
@ManasviGoyal I wanted to see the test failing so I am extending the index by 1 and content by 3, surprisingly the test passes anyway...do you see what is wrong?

ManasviGoyal replied:
@zonca That is strange. I checked and for me it failed.
```
void test_Indexed(): Assertion `builder.is_valid(error) == true'
```

zonca replied:
@ManasviGoyal what is worrysome is that also CI passed fine.
the only strangeness is that it prints out `Executing test_Indexed::endgroup::` but now sure what that means:

https://github.com/zonca/awkward/actions/runs/7103861309/job/19337516467?pr=3

zonca replied:
I figured out the issue, see https://github.com/scikit-hep/awkward/pull/2870

ManasviGoyal replied:
This makes sense. I was testing it by running the specific test file in C++ only that's why the test for failing for me.

zonca committed:
ci: compile in Debug mode

zonca committed:
test: fix unit test, check CI passes

zonca committed:
ci: deploy branch preview docs only on main repo

Forked repositories normally do not have secrets correctly configured
See
https://github.com/orgs/community/discussions/26409#discussioncomment-3251822

ManasviGoyal commented on header-only/tests/test_1494-layout-builder.cpp:
```diff
@@ -1209,6 +1212,62 @@ test_Regular_size0() {
+  IndexedBuilder<uint32_t, NumpyBuilder<double>> builder;
+  assert(builder.length() == 0);
+
+  std::cout << "Executing test_Indexed";
```
Please remove this.

ManasviGoyal commented on header-only/tests/test_1494-layout-builder.cpp:
```diff
@@ -1209,6 +1212,62 @@ test_Regular_size0() {
+  assert(builder.length() == 0);
+}
+
 void
 test_Indexed_as_IndexedOption() {
```
This test can be removed now.

zonca committed:
fix: remove leftover cout

zonca committed:
test: removed the test_Indexed_as_IndexedOption test

ManasviGoyal commented:
@zonca please make a PR to awkward if you are done with it. Thanks!

jpivarski reviewed (approved):
@ManasviGoyal has already approved this, and we only require one approval, but I looked it over also and approve it as well.

I'll merge it. Thanks, @zonca!

jpivarski reviewed (approved):
For some reason, that didn't set the "approve" checkmark. This should do it.

jpivarski commented:
Nope. Oh well!

jpivarski commented:
Wait a minute, why can't I merge this? Oh! It's a PR into zonca/awkward. I see scikit-hep/awkward#2883 and will go there now.

zonca commented:
I'm closing this PR, let's refer to the one in the official repo
